mitodl/mitxpro
# Added edX unenrollment capability

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes #1068 
Fixes #988 

#### What's this PR do?
Adds edX unenrollment capability and ensures that our management commands that deactivate course run enrollments also attempt to unenroll the given user in edX

#### How should this be manually tested?
The `transfer_enrollment` command is probably easiest for testing since it can be done multiple times in a row. As always, the enrolled course run must exist in edX, so for local testing the `course-v1:edX+DemoX+Demo_Course` run will probably be best to use.

1. Run devstack
1. Enroll a user in the course run 
1. Check that user's edX dashboard to make sure the course run is shown
1. Transfer the enrollment to another user: `manage.py transfer_enrollment --from-user "user1@example.com" --to-user "user2@example.com" --run "course-v1:edX+DemoX+Demo_Course"`
1. Check the first user's edX dashboard. The course run should now be gone
1. Log into xPro as the other user. That course run should be in their xPro and edX dashboards now

#### Any background context you want to provide?
- This PR partially address the issues we have with users having multiple enrollment records in the same course run (#1002). For our management commands, the logic is changed to just fetch the most recently created enrollment record
- I also noticed a minor bug with `AuditableModel` and fixed it in this PR. The `save_and_log` method used `<model>.objects` to fetch the original object data. `objects` can be overridden though, and we do exactly that for `CourseRunEnrollment`/`ProgramEnrollment`. In those cases, if we wanted to update a record set to `is_active=False` and log the changes, the wrong queryset would be used and it would fail.
